### DISCUSSION TITLE
Coefficient of restitution

### DISCUSSION THREAD

**Discussion by Aryan-Satpathy** [google-deepmind/mujoco#2081] [Asking for Help] [answered]
Hi, I am a student trying to use MuJoCo for some physics simulation. I was wondering if there is a way to define the coefficient of restitution for geometries, or somehow translate coefficient of restitution to solimp and/or solref parameters?

**Comment by kevinzakka** [comment]
https://mujoco.readthedocs.io/en/latest/modeling.html#restitution

**Reply by Aryan-Satpathy** [reply]
Hi Kevin, thank you for the pointer and apologies for the delayed response. I had gone through the documentation earlier. It does explain how to get a perfectly elastic collision. However I was wondering if there was a way to find the stiffness and damping parameters for a geometry, given an arbitrary coefficient of restitution e.

**Reply by JackSchoneveld** [reply]
HI Aryan, I am also working on simulations for MuJoCo regarding the restitution coefficient. Did you manage to find this relation?

**Reply by Aryan-Satpathy** [reply, marked as answer]
Hi Jack, I couldn't find a way to do it through MuJoCo. So, I approximated the collision as a damped spring between the bodies (when they are in contact) and manually applied the force before calling `mujoco.mj_step()`. The parameters are calculated as follows (this can be derived using energy conservation):

```
k = -c.solref[0]    # Set some high k value, I choose k = 2500
b = -2*math.sqrt(k*m_eff)*math.log(min(ce, 1e-15))/np.pi  # 1e-15 to avoid errors when ce = 0 (inelastic collision)
```
`ce` is the coefficient of restitution, and `m_eff` is the harmonic mean of the two masses that are colliding. You need to consider inertia matrices and jacobians for the colliding bodies if there are joints on any of the bodies or if they are not point masses. It works accurately enough in my tests.

**Comment by EmmanuelMess** [comment]
I used contact/pair to change the coefficient until it fit experiments, (for a bicycle wheel):
```
<mujoco>
  <option cone="elliptic" timestep="1e-4" integrator="implicitfast"/>

  <visual>
    <global elevation="0"/>
  </visual>

  <asset>
    <texture name="grid" type="2d" builtin="checker" rgb1=".1 .2 .3" rgb2=".2 .3 .4" width="300" height="300"/>
    <material name="grid" texture="grid" texrepeat="8 8" reflectance=".2"/>
    <texture name="ball" type="cube" builtin="checker" rgb1=".3 .3 .3" rgb2="1 1 1" width="300" height="300"/>
    <material name="ball" texture="ball" texrepeat="8 8"/>
  </asset>

  <worldbody>
    <geom name="floor" size="1 1 .01" type="plane" material="grid"/>
    <geom name="wall" size="1 1 .01" type="plane" material="grid" pos="1 0 1" zaxis="-1 0 0"/>
    <light pos="0 0 2"/>
    <light pos="-2 0 2" dir="1 0 -1"/>
    <body name="wheel" pos="-.1 0 1">
      <joint type="slide" axis="0 0 1"/>
      <geom name="wheel" type="cylinder" size="0.33" fromto="-0.025 0 0 0.025 0 0" mass="2.2"/>
    </body>
  </worldbody>
  
  <sensor>
    <framepos  objtype="xbody" objname="wheel" />
    <framelinvel  objtype="xbody" objname="wheel" />
  </sensor>
  
  <contact>
    <pair geom1="floor" geom2="wheel" solref="-10000 -20"/>
    <pair geom1="wall" geom2="wheel" solref="-10000 -20"/>
  </contact>

  <keyframe>
    <key name="throw" qvel="0"/>
  </keyframe>
</mujoco>
```
